### PR TITLE
hcm: log before onDestroy

### LINKED ISSUE #12616: log() called for filters registering via addAccessLogHandler after the filter chain has been destroyed
Description:

See the title.

The ActiveStream is being destroyed by a deferred delete and it is occurring after the filter chain  has been destroyed. It looks like the UpstreamRequest destructor is calling the upstream_logs_ ->log() call which in turn is calling the log() API in the filter for a filter chain which is already destroyed.  While it is easy enough to detect this and avoid using state which is no longer valid I am wondering if this is the intended design?  I would think that it would be best to avoid calls to filters on a filter chain which has already been destoryed.

[optional Relevant Links:]

https://github.com/envoyproxy/envoy-wasm/issues/615

This occurs during rapid reconfiguration of Wasm filters.

Repro steps:

Add a Wasm filter which which includes a log() handler (e.g. HTTP filter) and reconfigure the filter chains rapidly (every 2 seconds) while under load.

[2020-08-11 21:13:00.416][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Segmentation fault, suspect faulting address 0x560500000000
[2020-08-11 21:13:00.416][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-08-11 21:13:00.416][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: e275bca85d3412681b18275274d61c55573db731/1.15.0/Modified/DEBUG/BoringSSL
[2020-08-11 21:13:00.478][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x560523a9f33c]
[2020-08-11 21:13:00.479][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: __restore_rt [0x7f19533f68a0]
[2020-08-11 21:13:00.536][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: proxy_wasm::(anonymous namespace)::V8::getModuleFunctionImpl<>()::{lambda()#1}::operator()() [0x560521d9eca5]
[2020-08-11 21:13:00.594][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: std::_Function_handler<>::_M_invoke() [0x560521d9ea62]
[2020-08-11 21:13:00.652][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: std::function<>::operator()() [0x560521d79da0]
[2020-08-11 21:13:00.709][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: proxy_wasm::ContextBase::onLog() [0x560521d760bc]
[2020-08-11 21:13:00.767][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Extensions::Common::Wasm::Context::log() [0x560521a107d3]
[2020-08-11 21:13:00.825][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: Envoy::Http::ConnectionManagerImpl::ActiveStream::~ActiveStream() [0x56052358450d]
[2020-08-11 21:13:00.883][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: Envoy::Http::ConnectionManagerImpl::ActiveStream::~ActiveStream() [0x560523584e20]
[2020-08-11 21:13:00.941][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Http::ConnectionManagerImpl::ActiveStream::~ActiveStream() [0x560523584f19]
[2020-08-11 21:13:00.999][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: std::default_delete<>::operator()() [0x560520abb7df]
[2020-08-11 21:13:01.056][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #11: std::unique_ptr<>::reset() [0x560521a555c6]
[2020-08-11 21:13:01.114][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #12: Envoy::Event::DispatcherImpl::clearDeferredDeleteList() [0x560522fa4bb0]
[2020-08-11 21:13:01.173][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #13: Envoy::Event::DispatcherImpl::DispatcherImpl()::$_0::operator()() [0x560522fa70f9]
[2020-08-11 21:13:01.232][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #14: std::_Function_handler<>::_M_invoke() [0x560522fa6fad]
[2020-08-11 21:13:01.290][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #15: std::function<>::operator()() [0x56052052fbce]
[2020-08-11 21:13:01.347][26263][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #16: Envoy::Event::SchedulableCallbackImpl::SchedulableCallbackImpl()::$_0::operator()() [0x560522fdbcb5]

### LINKED ISSUE #12840: Reduce code duplication in test/common/http/conn_manager_impl_test.cc
The tests in this file have a lot of duplication (e.g. around resetting the stream at the end). It would be helpful to go through and extract this into helper functions to improve readability

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Commit Message: call log() before onDestroy(), as well as populate final stream info details at the time of log() rather than the deferred destruction.

Additional Description:
Risk Level: low (might affect timings of events)
Testing: unit
Docs Changes: none
Release Notes: none
Fixes #12616 
Fixes #12840

### CODE REVIEW

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -577,6 +580,14 @@ ConnectionManagerImpl::ActiveStream::ActiveStream(ConnectionManagerImpl& connect
 }
 
 ConnectionManagerImpl::ActiveStream::~ActiveStream() {
+  if (connection_manager_.codec_->protocol() < Protocol::Http2) {
+    // For HTTP/2 there are still some reset cases where details are not set.
+    // For HTTP/1 there shouldn't be any. Regression-proof this.
+    ASSERT(filter_manager_.streamInfo().responseCodeDetails().has_value());
```
Comment (kyessenov) [inline source/common/http/conn_manager_impl.cc:586]: @snowp @alyssawilk I suspect there is some race here with onEvent(RemoteClose) and deferred delete. We may not be able to detect remote close if we log immediately at the end of the stream. Any thoughts?

Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc:593]: If we're going to move logs around, think I'd like this assert removed to before we log (we want to make sure details are set before stream info is shipped to access logs)

I don't think we should have to set DownstreamRemoteDisconnect in two places, and I'm not sure what race there would be.  If we get a normal end_stream set, the stream should finish inline and not be lingering around to pick up and I'd think any subsequent disconnects wouldn't be registered?

Comment (kyessenov) [reply source/common/http/conn_manager_impl.cc:593]: I'm getting test failures if I don't update stream info in the destructor. Might be a test issue again. Is it possible to get onEvent(RemoteClose) once stream is scheduled to be defer-deleted? If not, then it is likely just another unrealistic test.

Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc:593]: yeah, I'd guess it's the tests being badly ordered but you should be able to check with gdb or envoy logs.

Comment (kyessenov) [reply source/common/http/conn_manager_impl.cc:593]: What I mean with the race is the following sequence:
1. Socket is read ready, data read, calling onData() and and encodeHeaders() in HCM.
2. encodeHeaders() ends stream, triggers log().
3. Socket is remote-closed.

It seems that we need to process the file events before triggering log() but I'm not sure how to hold onData() for tht.

Comment (alyssawilk) [reply source/common/http/conn_manager_impl.cc:593]: ah, in that case in production I think the details would basically be set on encodeHeaders (like the router will when proxying from upstream, or the HCM will when sendLocalReply is called) so maybe switch to that?

Comment (kyessenov) [reply source/common/http/conn_manager_impl.cc]: Thanks for explaining. That works indeed.

**Comment by kyessenov** [comment]
Windows failure looks like a flake in ext_authz test. Can someone re-kick windows CI?

**Comment by snowp** [comment]
/retest

should work for Azure pipelines now

**Comment by alyssawilk** [comment]
Oh actually already tagged for Snow, so I'll leave that as is :-P

**Pull request merged**